drivers/serial/uart_intel_lw.c

```diff
@@ -0,0 +1,999 @@
+ * stores data that cannot be changed during run time.
+ */
+struct uart_intel_lw_device_config {
+	mm_reg_t base;
+	uint32_t flags;              /* refer to configuration flags */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t  irq_config_func;
+	unsigned int irq_num;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+/**
+ * function prototypes
+ */
+static int uart_intel_lw_irq_update(const struct device *dev);
+static int uart_intel_lw_irq_tx_ready(const struct device *dev);
+static int uart_intel_lw_irq_rx_ready(const struct device *dev);
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+/**
+ * @brief Poll the device for input.
+ *
+ * This is a non-blocking function.
+ *
+ * This driver supports interrupt driven api.
+ * Polling for data under normal operation, might cause unexpected behaviour.
+ * If users wish to poll for data, please use the api:
+ * uart_drv_cmd with CMD_POLL_ASSERT_RTS before polling in.
+ * Then use CMD_POLL_DEASSERT_RTS to resume normal operation after polling.
+ *
+ * @param dev UART device instance
+ * @param p_char Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if input buffer is empty.
+ * -EINVAL if p_char is null pointer
+ */
+static int uart_intel_lw_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	const struct uart_intel_lw_device_config *config = dev->config;
+	struct uart_intel_lw_device_data *data = dev->data;
+	int ret_val = -1;
+	uint32_t status;
+
+	/* generate fatal error if CONFIG_ASSERT is enabled. */
+	__ASSERT(p_char != NULL, "p_char is null pointer!");
+
+	/* Stop, if p_char is null pointer */
+	if (p_char == NULL) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	/* check if received character is ready.*/
+	status = sys_read32(config->base + INTEL_LW_UART_STATUS_REG_OFFSET);
+	if (status & INTEL_LW_UART_STATUS_RRDY_MSK) {
+		/* got a character */
+		*p_char = sys_read32(config->base + INTEL_LW_UART_RXDATA_REG_OFFSET);
+		ret_val = 0;
+	}
+
+	k_spin_unlock(&data->lock, key);
+
+	return ret_val;
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * This function will block until transmitter is ready.
+ * Then, a character will be transmitted.
+ *
+ * This driver supports interrupt driven api.
+ * Polling for data under normal operation, might cause unexpected behaviour.
+ * If users wish to poll for data, please use the api:
+ * uart_drv_cmd with CMD_POLL_ASSERT_RTS before polling out.
+ * Then use CMD_POLL_DEASSERT_RTS to resume normal operation after polling.
+ * Please do not set CMD_TRBK_EN, when polling out data.
+ *
+ * @param dev UART device instance
+ * @param c Character to send
+ */
+static void uart_intel_lw_poll_out(const struct device *dev, unsigned char c)
+{
+	const struct uart_intel_lw_device_config *config = dev->config;
+	struct uart_intel_lw_device_data *data = dev->data;
+	uint32_t status;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
```

Comment:
Consider refactoring so the spinlock is acquired/released inside the loop to allow ISRs to occur promptly

Reply:
@gohshunjing doing blocking IO in a spinlock is not good practice, it prevents higher priority threads and ISRs from occurring.

I would suggest something like this:

    while (!done) {
        spin_lock();

        if (tx_avail()) {
            tx(c);
            done = true;
        }

        spin_unlock();
        
        if (!done) {
            yield();
        }
    }

> I have looked at other uart drivers, most of them have spinlock outside of the loop.
If this refactoring is required, should the other uart driver be updated as well?

Yes, I think those should be updated as well, although I am not a maintainer of that code

Reply:
Hi @gramsay0,

i have refactored the spinlocks to be outside of loop. Thank you for your feedback.

best regards,
shun jing

Reply:
@gohshunjing thanks, also it might be slightly better to assert/deassert RTS within the lock (I overlooked that when I first looked):

    spin_lock();

    while (true) {
        if (tx_avail()) {
            assert_rts();
            tx(c);
            deassert_rts();
            break;
        } else {
            /* Allow ISRs and same/higher priority threads some CPU time */
            spin_unlock();
            k_yield();
            spin_lock();
        }
    }

    spin_unlock();

Reply:
thank you for the feedback. I have updated the code.

Reply:
@gohshunjing when `CONFIG_UART_INTEL_LW_AUTO_LINE_CTRL_POLL` is enabled the `uart_intel_lw_poll_out` function will set/clear RTS on every loop while the TX FIFO is full.

I suggest moving the RTS code inside the `if (status & INTEL_LW_UART_STATUS_TRDY_MSK)` scope

Reply:
Hi @gramsay0, i have updated the code.
During the update, i have found a bug, where only first character is polled out, when CONFIG_UART_INTEL_LW_AUTO_LINE_CTRL_POLL is enabled and uart_poll_out is called consecutively in a loop.
This is because the RTS is already cleared, before the first byte in tx fifo is shifted out completely. When RTS is cleared, the HW will stop transmitting the other bytes. To solve this, i have changed the code to wait until the tx data is shifted out before RTS is cleared.
@dcpleung can you please review uart_intel_lw_poll_out again? Thank you.